---
title: "*Isoetes* RADSeq data analysis"
author: Sylvia Kinosian
output: html_document
---

Data processing and analysis for PNW Isoetes project with Jacob Suissa. Unless otherwise noted, all analysis were performed using Ubuntu 18.04 LTS. Please see the Program Versions tab for information on programs used.

# {.tabset}

## Demultiplexing

### Stacks

```{bash eval=FALSE}
~/apps/stacks-2.5/process_radtags -P -p /media/kaiser/skinosian/isoetes/raw/ -o ./ -b isoetes_barcodes.txt -c -q -r --inline_null --renz_1 pstI --renz_2 mspI


395070650 total sequences
  8656066 barcode not found drops (2.2%)
    20178 low quality read drops (0.0%)
  3701923 RAD cutsite not found drops (0.9%)
382692483 retained reads (96.9%)
```

### Stats

## Data Processing

We utilized the ipyrad data processing pipeline (Eaton & Overcast 2020)

```{bash eval=FALSE}
ipyrad -p params.txt -s 1234567
```

```{bash eval=FALSE}
------- ipyrad params file (v.0.7.30)-------------------------------------------
isoetes                        ## [0] [assembly_name]: Assembly name. Used to name output directories
                               ## [1] [project_dir]: Project dir (made in curdir if not present)
                               ## [2] [raw_fastq_path]: Location of raw non-demultiplexed fastq files
                               ## [3] [barcodes_path]: Location of barcodes file
./fastqs/*.fastq.gz            ## [4] [sorted_fastq_path]: Location of demultiplexed/sorted fastq files
denovo                         ## [5] [assembly_method]: Assembly method 
                               ## [6] [reference_sequence]: Location of reference sequence file
pairddrad                      ## [7] [datatype]: Datatype (see docs): rad, gbs, ddrad, etc.
TGCAG, GCC                     ## [8] [restriction_overhang]: Restriction overhang (cut1,) or (cut1, cut2)
5                              ## [9] [max_low_qual_bases]: Max low quality base calls (Q<20) in a read
33                             ## [10] [phred_Qscore_offset]: phred Q score offset (33 is default)
6                              ## [11] [mindepth_statistical]: Min depth for statistical base calling
6                              ## [12] [mindepth_majrule]: Min depth for majority-rule base calling
10000                          ## [13] [maxdepth]: Max cluster depth within samples
0.90                           ## [14] [clust_threshold]: Clustering threshold for de novo assembly
0                              ## [15] [max_barcode_mismatch]: Max number of allowable mismatches in barcodes
2                              ## [16] [filter_adapters]: Filter for adapters/primers (1 or 2=stricter)
35                             ## [17] [filter_min_trim_len]: Min length of reads after adapter trim
2                              ## [18] [max_alleles_consens]: Max alleles per site in consensus sequences
5, 5                           ## [19] [max_Ns_consens]: Max N's (uncalled bases) in consensus (R1, R2)
8, 8                           ## [20] [max_Hs_consens]: Max Hs (heterozygotes) in consensus (R1, R2)
20                             ## [21] [min_samples_locus]: Min # samples per locus for output
20, 20                         ## [22] [max_SNPs_locus]: Max # SNPs per locus (R1, R2)
8, 8                           ## [23] [max_Indels_locus]: Max # of indels per locus (R1, R2)
0.5                            ## [24] [max_shared_Hs_locus]: Max # heterozygous sites per locus (R1, R2)
0, 0, 0, 0                     ## [25] [trim_reads]: Trim raw read edges (R1>, <R1, R2>, <R2) (see docs)
0, 0, 0, 0                     ## [26] [trim_loci]: Trim locus edges (see docs) (R1>, <R1, R2>, <R2)
*                              ## [27] [output_formats]: Output formats (see docs)
                               ## [28] [pop_assign_file]: Path to population assignment file
```

#### Parameter Descriptions

Full documentation for all ipyrad parameters can be found [here](https://ipyrad.readthedocs.io/parameters.html). Below are descriptions of a few key parameters that we chose in our analysis.

[#5](https://ipyrad.readthedocs.io/parameters.html#assembly-method) **Assemby Method** - Since the available reference genomes for *Isoetes* are not very closely related to this clade, we decided to use the *de novo* assembly option for our ddRAD data. ipyrad offers four different assembly methods; for the [denovo method](https://ipyrad.readthedocs.io/methods.html#assembly-methods), raw sequences are assembled without a reference; homology is inferred using [vsearch](https://github.com/torognes/vsearch)
<br>
[#8](https://ipyrad.readthedocs.io/parameters.html#restriction-overhang) **Restriction Overhang** - We used the enzymes PstI and MspI for our library preparation. [PstI](https://en.wikipedia.org/wiki/PstI) cuts at the 3' end, so you need to reverse compliment of the overhang: TGCAG. [MspI](https://www.thermofisher.com/order/catalog/product/ER0541#/ER0541) cuts at the 5' end, so is: GCC.
<br>
[#16](https://ipyrad.readthedocs.io/parameters.html#filter-adapters) **Filter Adaptors** - We chose the most strict filtering option here, to remove not only barcodes, but Illumina and cutsite adaptors as well. During Step #2, reads are searched for the common Illumina adapter, plus the reverse complement of the second cut site (if present), plus the barcode (if present), and this part of the read is trimmed. 
<br>
[#21](https://ipyrad.readthedocs.io/parameters.html#min-samples-locus) **Min Samples per Locus** - This parameter sets the minimum number of samples that must have data for a given locus in the final data output. We chose to go with a relatively high number of minimum samples - 20 or a bit less than half of the total samples. We wanted our final output file to have data from most of our individuals.


## Data Analysis

### Population structure

To run STRUCTURE, we used the Center for High Performance Computing at the University of Utah ([CHPC](https://it.utah.edu/departments/chpc/)).

We ran 50 chains each of *K* = 2-10.

```{bash eval=FALSE}
for k in {2..10};
do
	for r in {1..50);
	do
		structure -i iso.str -m mainparams -e extraparams -K $k -o out_iso_$k-$r &
		sleep 3s
	done
done	
```

To combine chains across *Ks* we used the [CLUMMPAK server](http://clumpak.tau.ac.il/). We also used CLUMMPAK to estimate the Best K from Evanno et al. 2005

```{r eval=FALSE}
# load in ks from clummpack files
k2 <- read.csv("k2.txt", sep = '', header = F)
k2 <- k2[,-(1:5)]
k3 <- read.csv("k3.txt", sep = '', header = F)
k3 <- k3[,-(1:5)]
k4 <- read.csv("k4.txt", sep = '', header = F)
k4 <- k4[,-(1:5)]
k5 <- read.csv("k5.txt", sep = '', header = F)
k5 <- k5[,-(1:5)]

# names file includes individual ids, species names, and geographic locations
names <- read.csv("names.csv", sep = ',', header = T)

x <- as.data.frame(matrix(ncol = 18, nrow = 50))
x[,1:2] <- k2
x[,3:5] <- k3
x[,6:9] <- k4
x[,10:14] <- k5
#x[,15:20] <- k5
x[,15:18] <- names

# order by species, then geography
x <- x[order(x[,17]),]

# list for plotting
klist <- list(x[,1:2], x[,3:5], x[,6:9], x[,10:14])

structure_plot(x[,18], ninds=50, klist)

#######################################################
# functions needed
######################################################

# plotting and labeling function
structure_plot <- function(labels, ninds = 47, klist){
    # define colors
    cols <- c('#A8FFFD', '#A39D9D','#FFFF00', '#ff5a5a', '#69C261', '#26CDCD', '#B862D3','#C1C6FF')
    # unique label names
    sp.names <- as.character(unique(labels))
    #n <- as.data.frame(matrix(ncol = 1, nrow = ninds))
    #n[,1] <- names
    # locations of each column
    b <- as.data.frame(matrix(ncol = 1, nrow = ninds))
    b[,1] <- barplot(t(klist[[1]][1]), beside= F, col= cols, cex.name= 1, cex.axis= 1.2, border = 1, space = 0.05, xaxt = 'n', yaxt = 'n', cex.lab = 1, cex.main = 2)
    # find locations for labels in the barplot
    my.mean <- tapply(X = b[,1], INDEX = labels, mean)
    my.min <- tapply(X = b[,1], INDEX = labels, min)
    my.max <- tapply(X = b[,1], INDEX = labels, max)
    # data frame for plotting
    d <- sp_labels(names = sp.names, min = my.min, mean = my.mean, max = my.max)
    # plot
    plot_q_per_chain(klist, labels)
    #text(cex = 1.3, x = (d[,2]-0.3), y = -0.7, labels = d[,1], xpd=NA, srt=50, font=3)
    # lines
    #for (i in 1:length(d[,1])){
    #    lines(x = d[i,3:4] , y = rep(-0.1, 2), lwd = 2.5, col = "black", xpd = NA)
    #}

}

# create labels
sp_labels <- function(names, min, mean, max, ...){
    d <- as.data.frame(matrix(nrow = length(names), ncol = 4))
    for (j in 1:length(names)){
            d[j,1] <- names[j]
            d[j,3] <- min[[j]][1]
            d[j,2] <- mean[[j]][1]
            d[j,4] <- max[[j]][1]
    }
    return(d)
}

# plot chains with species and geography labels 
#plot_q_per_chain <- function(kqlist, ...){
#    # thalictroides 2, gaudichaudii, thalictroides 1, pteridoides, cornuta, misc, 
#    cols <- c('#A8FFFD','#B862D3', '#A39D9D','#FFFF00', '#69C261', '#FF59AC', '#26CDCD',  '#C1C6FF') 
#    #cols <- c('#000075', '#E6194B', '#AAFFC3', '#FFE119', '#F58231', '#3CB44B')
#    par(mfrow = c(length(kqlist),1), mar = c(1,3,3,1) + 0.1, oma = c(15,0,0,0), mgp = c(1,1,0))
#    chain <- seq(1, length(kqlist), 1) 
#    for(i in 1:length(kqlist)){
#        barplot(t(kqlist[[i]]), beside= F, col= cols, border = 1, space = 0.05, xaxt = 'n', yaxt = 'n', main = paste("k =", chain[i]+1, sep = ' '), cex.lab = 1.2, cex.main = 1.6)
#        # y axis
#        axis(2, at = c(0, 0.25, 0.5, 0.75, 1), cex.axis = 1, las = 2, pos = -0.2)
#    }
#}

##################################################
# different colors

plot_q_per_chain <- function(kqlist, xlabel){
	# brown, green, blue, grey 
	cols <- c('#A8886C', '#95CD85', '#4EC5FF', '#DBF3FF', '#A39D9D') 
	par(mfrow = c(length(kqlist), 1), mar = c(1, 3, 3, 1) + 0.1, oma = c(15, 0, 0, 0), mgp = c(1, 1, 0))
	chain <- seq(1, length(kqlist), 1) 
	for(i in 1){
        	barplot(t(kqlist[[i]]), beside = F, col = cols, border = 1, space = 0.05, xaxt = 'n', yaxt = 'n', main = paste("k =", chain[i]+1, sep = ' '), cex.lab = 1.2, cex.main = 1.6)
		#barplot(t(kqlist[[i]]), beside= F, col= cols, las= 2, axisnames= T, cex.name= 1, cex.axis= 1.2, border= 1.5, space= c(0.05,0.05), yaxt= 'n', ylab= paste("k =", chain[i]+1, sep= ' '), cex.lab= 2, names.arg= xlabel)

        	# y axis
        	axis(2, at = c(0, 0.5, 1), cex.axis = 1, las = 2, pos = -0.2)
    }
	barplot(t(kqlist[[length(kqlist)]]), beside = F, col = cols, border = 1, space = 0.05, yaxt = 'n', main = paste("k =", length(kqlist)+1, sep= ' '), cex.lab= 0.2, cex.main = 1, cex.names = 0.5, names.arg = xlabel, las = 2)
	axis(2, at = c(0, 0.5, 1), cex.axis = 1, las = 2, pos = -0.2)

}

plot_q_per_chain(k4list, ck4[,5])


```


## Program Versions

Below is a list of all program versions used in this analysis. **Please note** that newer versions of these software packages *may* work for this pipeline, but be aware that usage often changes with new versions. 

[stacks v 2.5](http://catchenlab.life.illinois.edu/stacks/)

[ipyrad](https://ipyrad.readthedocs.io/) [release: 0.7.30](https://github.com/dereneaton/ipyrad/releases/tag/0.7.30)

[STRUCTURE v.2.3.4](https://web.stanford.edu/group/pritchardlab/structure_software/release_versions/v2.3.4/html/structure.html)

[Cluster Markov Packager Across K (CLUMMPAK)](http://clumpak.tau.ac.il/) 

[Perl 5](https://www.perl.org/)

[Python 2.7.13](https://www.python.org/downloads/release/python-2713/)

[R v. >3.4](https://www.r-project.org/)
